### PR TITLE
Caching: Consolidate resource cache checking and updating in plugin middleware

### LINKED ISSUE #66297: Caching: Move call to update resource cache from plugin_resource.go to the caching middleware
In converting the query caching middleware to a service, we took a small shortcut that separates the resource cache retrieval from the resource cache update, making the code hard to follow. 

Marcus pointed us to an example that intercepts resource stream messages and re-sends them down the pipeline: https://github.com/grafana/grafana/pull/65616#discussion_r1159976153

We could do something similar in the caching middleware to intercept and cache the resources from the middleware itself.

### PR DESCRIPTION
**Pull request by mmandrus**
**What is this feature?**

This resolves [tech debt](https://github.com/grafana/grafana/issues/66297) in the Query Caching Service by having resource caching mirror query caching's cache updating implementation, and having the resource cache check + update occur in the same place.

It should also have the side effect of fixing https://github.com/grafana/grafana-enterprise/issues/4922, because the enterprise implementation won't read and write the response headers simultaneously.

**Why do we need this feature?**

Resolves tech debt and fixes a bug.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-enterprise/issues/4922
Fixes https://github.com/grafana/grafana/issues/66297

**Special notes for your reviewer:**

Enterprise PR https://github.com/grafana/grafana-enterprise/pull/4969

### CODE REVIEW

**Review by marefr — APPROVED** [review]
LGTM great stuff :+1: 

My only question is related to the following response header logic:
https://github.com/grafana/grafana/blob/6abd75b69cb967b6cfca3351c6a3bcfa3dd94935/pkg/api/plugin_resource.go#L191-L194

Are you somehow dependent on the response headers set in plugin_resource.go? My understanding is it shouldn't matter since serving a cached response would go through that logic and AFAIK apply the response header logic even if it's cached or not, correct?

**Comment by mmandrus** [comment]
> LGTM great stuff 👍
> 
> My only question is related to the following response header logic:
> 
> https://github.com/grafana/grafana/blob/6abd75b69cb967b6cfca3351c6a3bcfa3dd94935/pkg/api/plugin_resource.go#L191-L194
> 
> Are you somehow dependent on the response headers set in plugin_resource.go? My understanding is it shouldn't matter since serving a cached response would go through that logic and AFAIK apply the response header logic even if it's cached or not, correct?

The old caching implementation (using middleware, skipping the plugin_resource.go logic entirely) actually did the same thing: perform all of the queries (either to the cache or to the plugin), then apply the content-type header as the final step before returning the response.

So, with the new workflow, we now get the same behavior as before :+1::skin-tone-2: We have no dependency on the content-type field in the caching impl itself.
